### ISSUE TITLE
Details about IR/RGB conversion

### ISSUE THREAD

**Issue by xelmirage** [mingyuliutw/UNIT#12]
Hi, Mingyu. 
I'm studying your amazing work about UNIT, and I'm very interested in UNIT's 
 attractive poteintial of IR/RGB conversion. So I tried to repeat the experiment. Since the detailed parameters of training are not provided, I just tried borrowing settings in synthia2cityscape.yaml for the training. But the results were not satisfying.
So I'm wondering is it right for me to do so? If it's not right to borrow settings in synthia2cityscape.yaml, is there any possibility that settings for IR/RGB conversion may be released?
Thanks for your attention and amazing work:)

**Comment by mingyuliutw** [comment]
@xelmirage I haven't tried IR/RGB conversion with this implementation and do not have an yaml file to share. But I am wondering about your setting. 

1> How many IR and RGB images do you have? The fundamental assumption of the work is that there exists a joint distribution and we have a reasonable amount of images to approximate the joint distribution. If these do not hold, then the algorithm may fail.

2> IR and RGB images have different number of channels and different range values. Their reconstruction losses may not be comparable. You may need to balance them by playing with the parameter. 

3> How many iterations have you trained. It takes several days in my computer to start seeing some reasonable results. 

Finally, it would be great if you could share some images if possible.

**Comment by xelmirage** [comment]
@mingyuliutw Thank you for your attention. 
1>At the beginning I wanted to do an exact repeat the IR/RGB conversion in the first edition of your paper, in which you mentioned that 31386 RGB images and 31191 IR images from KAIST dataset were used to train the Net. I took a wild guess that you may used set 2,5,7,10,11 to provide RGB images and set 1,3,6,9 to provide LWIR images. But set 5,10,11 are nighttime datasets, in my understanding, all the RGB images should be daytime images, so I used RGB images in set 1,2,7,8 instead of set 2,5,7,10,11.
2>I directly used the settings from synthia2cityscape.yaml, because in all the yamls provided, I think this conversion is the most similar one to IR/RGB conversion. The modified yaml is attached. The IR images are treated as RGB images with three same channels.
[kaist.zip](https://github.com/mingyuliutw/UNIT/files/1384693/kaist.zip)


3>The training has about 250000 iterations so far and is not showing satisfying results. The gen_total_loss, dis_true_acc_0, and dis_fake_acc_0 are shown in the figures below:
![image](https://user-images.githubusercontent.com/2993178/31575999-14c88588-b125-11e7-9a17-688081730f1d.png)
![image](https://user-images.githubusercontent.com/2993178/31576070-4d465682-b126-11e7-86ee-ce1d3c50cb77.png)
![image](https://user-images.githubusercontent.com/2993178/31576072-63a60602-b126-11e7-9672-714214e5de10.png)

The gen_total_loss increased after 8k iterations. because I'm not experienced in training networks, I don't know whether it's normal, in the training of CelebA dataset this also happens. So far the Net doesn't have a meaningful result, here are some results:
247500
![gen_00247500](https://user-images.githubusercontent.com/2993178/31576080-b0cf3e80-b126-11e7-96fd-84fc0946ad1b.jpg)


225000
![gen_00225000](https://user-images.githubusercontent.com/2993178/31576084-cde17326-b126-11e7-89eb-d38fd3b43cd0.jpg)


210000
![gen_00210000](https://user-images.githubusercontent.com/2993178/31576088-e1d14b0e-b126-11e7-9f77-b750f4f055fc.jpg)


165000
![gen_00165000](https://user-images.githubusercontent.com/2993178/31576092-f7574a32-b126-11e7-8fe9-ec10cf5e5644.jpg)


130000
![gen_00130000](https://user-images.githubusercontent.com/2993178/31576095-056cea32-b127-11e7-8fd0-c10e0a1da01a.jpg)


1000000
![gen_00100000](https://user-images.githubusercontent.com/2993178/31576097-17571ca4-b127-11e7-83db-26f0e2afe646.jpg)


85000
![gen_00085000](https://user-images.githubusercontent.com/2993178/31576110-466f7392-b127-11e7-875e-b53f7c3e97ed.jpg)



65000
![gen_00065000](https://user-images.githubusercontent.com/2993178/31576111-506c88d0-b127-11e7-86c2-5d2ec7858efd.jpg)



30000
![gen_00030000](https://user-images.githubusercontent.com/2993178/31576107-3c950e36-b127-11e7-95ce-9cb3eadac84d.jpg)


10000
![gen_00010000](https://user-images.githubusercontent.com/2993178/31576120-6131d8c8-b127-11e7-838d-154ee3dff825.jpg)


2500
![gen_00002500](https://user-images.githubusercontent.com/2993178/31576123-6a60fd16-b127-11e7-9644-77aba98b3025.jpg)



Thank you again for you precious time!

**Comment by mingyuliutw** [comment]
BTW, in the first version of the work, we used a context image idea. Basically, the input image is RGB+normalized y-coordinate. We also used the UNet architecture. I am not sure if this could be the difference. But these are something that I could thought of.

**Comment by xjoshramos** [comment]
can you share the trained weights for this conversion please? thank you

**Comment by xelmirage** [comment]
Hi @mingyuliutw , I have tried the version01 of UNIT. Since the v01 branch doesn't have a yaml for TIR-to-RGB conversion either, I modified the kaist_day_and_night.yaml. The modifications are replacing the class_name for TIR image to "dataset_thermal_image" as luckily there is a class for TIR images in the dataset_image.py, and the channel for TIR images is set to 1. However, there is also no satisyfing results after a really long training. So I have following thoughts:
1. As @xjoshramos commented, is there any pre-trained weights for TIR-RGB conversion available for evaluating the potential of UNIT?
2. Does the selection of training images have effects on the results? Because KAIST dataset has 12 sets and each set also have several subsets, Is it possible for me to know what sets and subsets (such as setXX-V00X) are used in training?
3.Is the  kaist_day_and_night.yaml has the correct settings for TIR-RGB conversion? If not, which item needs to be modified?

Thank you for your attention!

Some results of training using the version01 code:
gen_00055000
![gen_00055000](https://user-images.githubusercontent.com/2993178/33328793-e6c08354-d495-11e7-92dd-09a83aacf74b.jpg)
gen_00112500
![gen_00112500](https://user-images.githubusercontent.com/2993178/33328502-1e364a68-d495-11e7-96e7-ab8902305cd9.jpg)
gen_00150000
![gen_00150000](https://user-images.githubusercontent.com/2993178/33328812-f4751d84-d495-11e7-90fd-a03bc5adbeb5.jpg)
gen_00200000
![gen_00200000](https://user-images.githubusercontent.com/2993178/33328713-a368e0c4-d495-11e7-9c52-70129aa43d0a.jpg)

**Comment by mingyuliutw** [comment]
@xelmirage 

If you check the first version of the paper (I think you can find it in the arXiv website), you would see that in the implementation section, there is a paragraph about using context. Basically, we added one channel to both the RGB (3 channels to 4 channels) and thermal ( 1 channel to 2 channels) images where the additional channel is the normalized y-coordinates [0 1] of the pixels. I pointed out this in the early reply. I don't know if you noticed it. As I recall the version 1 code in the branch does not have this context image component implemented. You might need to implement the component to get better results. 

@xelmirage and @xjoshramos 

I do not have any pre-trained models (The true first version of the code was implemented in Caffe because PyTorch was not yet born. I reimplemented the algorithm using PyTorch and release the code for sharing.) from the first version anymore as I have shifted to my latest version and storing these models takes a lot of disk space.

**Comment by mingyuliutw** [comment]
BTW, if the goal is to make a comparison to the UNIT algorithm. I would recommend that you use the Cityscape and SYNTHIA datasets as images in the two datasets have high quality.

**Comment by xjoshramos** [comment]
@mingyuliutw Is this the paper?
https://arxiv.org/pdf/1703.00848v1.pdf

**Comment by mingyuliutw** [comment]
Yes

**Comment by xelmirage** [comment]
@mingyuliutw Thank you for your reply, as you first pointed out the context image component in your former replies, I checked the version01 and founded that version01 did have context image component. The related codes exist in dataset_image.py at line 26-34, and the yaml file also has context option switched to 1. So I'm pretty sure that I used the context image component as you mentioned.

**Comment by mingyuliutw** [comment]
@xelmirage 
I see. I forgot I reimplemented the context part. I guess in this case you need to explore various network design choices or various GAN training tricks.

**Comment by xelmirage** [comment]
@mingyuliutw Thank you very much for your help. Just one minor question, I am still wondering the details about the tranning dataset. Which part of KAIST images were chosen to make the 31386 IR images and 31191 RGB images... is this detail convenient to share?

**Comment by groot-1313** [comment]
@xelmirage , did  you obtain any good results of the colourization from thermal images? If yes, can you please share how to reproduce your results!

**Comment by mingyuliutw** [comment]
Just to point there was a bug in the xy image creation as described in [issues 39](https://github.com/mingyuliutw/UNIT/issues/39).

**Comment by groot-1313** [comment]
I see that the issue#39 is related to the context image component which was not part of the master branch at the time this issue was created. Does it mean I do not need to revert to version_01 for the image context part, and that the master branch would work just fine?

**Comment by mingyuliutw** [comment]
Sorry for the confusion. What’s I meant is the xy-image supposed to be useful for IR2RGB but it was not the case. I suspected it was due to the bug. That’s all. I will leave version one as it is.

**Comment by lghasemzadeh** [comment]
Hello,
I have the same concern, I need to convert IR images to RGB.
Have you achieved acceptable result? @xelmirage 
I appreciate if you share your experience about this issue.
Thx
